pkg/readiness/ready_tracker.go

```diff
@@ -689,53 +732,79 @@ func (t *Tracker) trackConstraintTemplates(ctx context.Context) error {
+// and any SyncSet resources present on the cluster.
+// Works best effort and fails-open if the a resource cannot be fetched or does not exist.
+func (t *Tracker) trackSyncSources(ctx context.Context) error {
 	defer func() {
-		defer t.config.ExpectationsDone()
-		log.V(1).Info("config expectations populated")
+		t.config.ExpectationsDone()
+		log.V(logging.DebugLevel).Info("config expectations populated")
 
-		wg.Wait()
+		t.syncsets.ExpectationsDone()
+		log.V(logging.DebugLevel).Info("syncset expectations populated")
 	}()
 
+	handled := make(map[schema.GroupVersionKind]struct{})
+
 	cfg, err := t.getConfigResource(ctx)
 	if err != nil {
-		return fmt.Errorf("fetching config resource: %w", err)
+		log.Error(err, "fetching config resource")
 	}
 	if cfg == nil {
 		log.Info("config resource not found - skipping for readiness")
-		return nil
-	}
-	if !cfg.GetDeletionTimestamp().IsZero() {
-		log.Info("config resource is being deleted - skipping for readiness")
-		return nil
+	} else {
+		if !cfg.GetDeletionTimestamp().IsZero() {
+			log.Info("config resource is being deleted - skipping for readiness")
+		} else {
+			t.config.Expect(cfg)
+			log.V(logging.DebugLevel).Info("setting expectations for config", "configCount", 1)
+
+			for _, entry := range cfg.Spec.Sync.SyncOnly {
+				handled[schema.GroupVersionKind{
+					Group:   entry.Group,
+					Version: entry.Version,
+					Kind:    entry.Kind,
+				}] = struct{}{}
+			}
+		}
 	}
 
 	if operations.HasValidationOperations() {
-		// Expect the resource kinds specified in the Config.
-		// We will fail-open (resolve expectations) for GVKs
-		// that are unregistered.
-		for _, entry := range cfg.Spec.Sync.SyncOnly {
-			gvk := schema.GroupVersionKind{
-				Group:   entry.Group,
-				Version: entry.Version,
-				Kind:    entry.Kind,
+		syncsets := &syncsetv1alpha1.SyncSetList{}
+		lister := retryLister(t.lister, retryAll)
+		if err := lister.List(ctx, syncsets); err != nil {
+			log.Error(err, "listing syncsets")
+		} else {
+			log.V(logging.DebugLevel).Info("setting expectations for syncsets", "syncsetCount", len(syncsets.Items))
+
+			for i := range syncsets.Items {
+				syncset := syncsets.Items[i]
+
+				t.syncsets.Expect(&syncset)
+				log.V(logging.DebugLevel).Info("expecting syncset", "name", syncset.GetName(), "namespace", syncset.GetNamespace())
+
+				for i := range syncset.Spec.GVKs {
+					gvk := syncset.Spec.GVKs[i].ToGroupVersionKind()
+					if _, ok := handled[gvk]; ok {
+						log.Info("duplicate GVK to sync", "gvk", gvk)
+					}
+
+					handled[gvk] = struct{}{}
+				}
 			}
-			u := &unstructured.Unstructured{}
-			u.SetGroupVersionKind(gvk)
-			t.config.Expect(u)
-			t.config.Observe(u) // we only care about the gvk entry in kinds()
+		}
+
+		// Expect the resource kinds specified in the Config resource and all SyncSet resources.
+		// We will fail-open (resolve expectations) for GVKs that are unregistered.
+		for gvk := range handled {
+			g := gvk
```

Comment:
why the extra variable?

Reply:
I think this is the classic go style data race when using a go func (closure) -- more info https://stackoverflow.com/questions/61641052/go-vet-loop-variable-i-captured-by-func-literal

tl;dr -- for loop "writes", go routine "reads", no synchronization, data race

Reply:
This is a side effect of `range`. It overwrites the variable, treating it like a pointer, g := gvk essentially makes a deepcopy